data-prepper-plugins/opensearch/src/main/java/org/opensearch/dataprepper/plugins/sink/opensearch/OpenSearchSink.java

```diff
@@ -201,11 +213,16 @@ public boolean isReady() {
 
   @Override
   public void doOutput(final Collection<Record<Event>> records) {
-    if (records.isEmpty()) {
-      return;
+    final long threadId = Thread.currentThread().getId();
+    if (!bulkRequestMap.containsKey(threadId)) {
+      bulkRequestMap.put(threadId, bulkRequestSupplier.get());
```

Comment:
Not a good idea. We do `bulkRequestSupplier.get()` in BulkRetryStrategy.java which may overwrite this. In general, this kind of logic could result in NullPointerExceptions as we have seen before.

Reply:
> We do bulkRequestSupplier.get() in BulkRetryStrategy.java which may overwrite this
Looking at the code in `BulkRetryStrategy`, a new BulkRequest is retrieved when there are errors and we need to retry. The BulkRequest is populated from the failed entries in the original bulk request then retried until success or retry limit reached.

This does not overwrite the bulkRequestMap entries here and also does not affect the logic to pack across batches since the `BulkRetryStrategy` is only invoked once a bulk request is ready to be sent. The `BulkRetryStrategy` handles the current BulkRequest to a completed state as well, the logic is never passed back to this class.

Please let me know if I am missing something.

> In general, this kind of logic could result in NullPointerExceptions as we have seen before.
Could you elaborate on how this would lead to an NPE? The entries in the map are tied to thread IDs, there is no way for a thread to interfere with another thread's entries in this map. 

There's no path to a missing key causing an NPE since we check that before getting the entry for a thread ID. We aren't removing entries from this map at any point so once the key is created it exists until the pipeline is stopped. 

There's no path to a value in this map being null since we are making the put calls against the supplier or an already created AccumulatingBulkRequest.

I understand there were issues with NPEs in a different part of the code that had multiple threads operating on the same map entry. Using the thread ID as the key is intended to prevent this as there are never threads operating on the same map entry.

Reply:
Also discussed on phone.